README.md
# Yabeda::Shoryuken

[![Gem Version](https://badge.fury.io/rb/yabeda-shoryuken.svg)](https://rubygems.org/gems/yabeda-shoryuken)

Yabeda plugin for easy [Shoryuken] gem collect metrics from your application

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'yabeda-shoryuken'
# Then add monitoring system adapter, e.g.:
# gem 'yabeda-cloudwatch'
```

And then execute:

    $ bundle

If you're not on Rails then configure Yabeda after your application was initialized:

```ruby
Yabeda.configure!
```

_If you're using Ruby on Rails then Yabeda will configure itself automatically!_

**And that is it!** Shoryuken metrics are being collected!

## Metrics

### Local per-process metrics

Metrics representing state of current Shoryuken worker process and stats of executed or executing jobs:

- Total number of executed jobs: `shoryuken_jobs_executed_total` -  (segmented by queue and class name)
- Number of jobs have been finished successfully: `shoryuken_jobs_success_total` (segmented by queue and class name)
- Number of jobs have been failed: `shoryuken_jobs_failed_total` (segmented by queue and class name)
- Time of job run: `shoryuken_job_runtime` (seconds per job execution, segmented by queue and class name)
- Maximum runtime of currently executing jobs: `shoryuken_running_job_runtime` (useful for detection of hung jobs, segmented by queue and class name)

### Client metrics

Metrics collected where jobs are being pushed to queues (everywhere):

- Total number of enqueued messages: `shoryuken_messages_enqueued_total_count` (segmented by `queue`)

## Custom tags

You can add additional tags to these metrics by declaring `yabeda_tags` method in your worker.

```ruby
# This block is optional but some adapters (like Prometheus) requires that all tags should be declared in advance
Yabeda.configure do
  default_tag :importance, nil
end
class MyWorker
  include Shoryuken::Worker
  def yabeda_tags(*params) # This method will be called first, before +perform+
    { importance: extract_importance(params) }
  end
  def perform(*params)
    # Your logic here
  end
end
```
## Configuration
Configuration is handled by [anyway_config] gem. With it you can load settings from environment variables (upcased and prefixed with `YABEDA_SIDEKIQ_`), YAML files, and other sources. See [anyway_config] docs for details.
Config key                | Type     | Default                                                 | Description                                                                                                                                        |
------------------------- | -------- | ------------------------------------------------------- |----------------------------------------------------------------------------------------------------------------------------------------------------|
`collect_cluster_metrics` | boolean  | Enabled in Shoryuken worker processes, disabled otherwise | Defines whether this Ruby process should collect and expose metrics representing state of the whole Shoryuken installation (queues, processes, etc). |
`declare_process_metrics` | boolean  | Enabled in Shoryuken worker processes, disabled otherwise | Declare metrics that are only tracked inside worker process even outside of them. Useful for multiprocess metric collection.                       |


## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/yabeda-rb/yabeda-shoryuken.

### Releasing

1. Bump version number in `lib/yabeda/shoryuken/version.rb`

   In case of pre-releases keep in mind [rubygems/rubygems#3086](https://github.com/rubygems/rubygems/issues/3086) and check version with command like `Gem::Version.new(Yabeda::Shoryuken::VERSION).to_s`

2. Fill `CHANGELOG.md` with missing changes, add header with version and date.

3. Make a commit:

   ```sh
   git add lib/yabeda/shoryuken/version.rb CHANGELOG.md
   version=$(ruby -r ./lib/yabeda/shoryuken/version.rb -e "puts Gem::Version.new(Yabeda::Shoryuken::VERSION)")
   git commit --message="${version}: " --edit
   ```

4. Create annotated tag:

   ```sh
   git tag v${version} --annotate --message="${version}: " --edit --sign
   ```

5. Fill version name into subject line and (optionally) some description (list of changes will be taken from changelog and appended automatically)

6. Push it:

   ```sh
   git push --follow-tags
   ```

7. GitHub Actions will create a new release, build and push gem into RubyGems! You're done!

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

[Shoryuken]: https://github.com/ruby-shoryuken/shoryuken "A super efficient Amazon SQS thread based message processor for Ruby"
[yabeda]: https://github.com/yabeda-rb/yabeda
[yabeda-prometheus]: https://github.com/yabeda-rb/yabeda-prometheus
[anyway_config]: https://github.com/palkan/anyway_config "Configuration library for Ruby gems and applications"

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 0.1.2 - 2023-01-31

- Removed unused dependency gems from gemspec. @retsef

## 0.1.1 - 2023-01-30

- Fix Shoryuken::ClientMiddleware implementation. @retsef
- Add missing test. @retsef

## 0.1.0 - 2023-01-25

- Add basic usage of the gem in README. @retsef
- Initial release of yabeda-shoryuken gem. @retsef

[@retsef]: https://github.com/retsef "Roberto Scinocca"

spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
require 'shoryuken'

# Fake that we're a worker to test worker-specific things
module Shoryuken::CLI; end

require 'yabeda/shoryuken'
require 'yabeda/rspec'

require_relative 'support/custom_metrics'
require_relative 'support/jobs'
require_relative 'support/shoryuken_inline_middlewares'
require_relative 'support/shoryuken_inline_processor'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.mock_with :rspec

  Kernel.srand config.seed
  config.order = :random

  config.before(:all) do
    Yabeda::Shoryuken.config.declare_process_metrics = true
    Yabeda::Shoryuken.config.collect_cluster_metrics = true

    Shoryuken.options[:concurrency] = 1
    Shoryuken.options[:delay]       = 1.0
    Shoryuken.options[:timeout]     = 1
    Shoryuken.options[:daemon]      = nil
    Shoryuken.options[:logfile]     = nil
    Shoryuken.options[:queues]      = nil

    Shoryuken.active_job_queue_name_prefixing = false
    Aws.config[:stub_responses] = true

    Shoryuken.sqs_client_receive_message_opts.clear
    Shoryuken.cache_visibility_timeout = false

    Shoryuken.worker_executor = Shoryuken::Worker::TestInlineExecutor
  end

  config.before do
    Shoryuken.sqs_client = instance_double('Aws::Sqs::Client').tap do |client|
      %w[default sample_plain sample_long_running sample_complex failing_plain].each do |queue_name|
        allow(client).to receive(:get_queue_url)
          .with({ queue_name: queue_name })
          .and_return instance_double('Aws::SQS::Types::GetQueueUrlResult', queue_url: queue_name)

        allow(client).to receive(:get_queue_attributes)
          .with({ attribute_names: ['All'], queue_url: queue_name })
          .and_return instance_double('Aws::SQS::Types::GetQueueAttributesResult', attributes: {})
      end

      allow(client).to receive(:send_message)
      allow(client).to receive(:receive_message)
    end
  end

  config.after(:all) do
    Shoryuken.worker_executor = Shoryuken::Worker::DefaultExecutor
  end
end


spec/support/jobs.rb
# frozen_string_literal: true

class DefaultJob
  include Shoryuken::Worker
  shoryuken_options queue: 'default'

  def perform(*_args)
    'My default job'
  end
end

class SamplePlainJob
  include Shoryuken::Worker
  shoryuken_options queue: 'sample_plain'

  def perform(*_args)
    'My job is simple'
  end
end

class SampleLongRunningJob
  include Shoryuken::Worker
  shoryuken_options queue: 'sample_long_running'

  def perform(*_args)
    sleep 0.05
    'Phew, I\'m done!'
  end
end

class SampleComplexJob
  include Shoryuken::Worker
  shoryuken_options queue: 'sample_complex'

  def perform(*_args)
    Yabeda.test.whatever.increment({ explicit: true })
    'My job is complex'
  end

  def yabeda_tags
    { implicit: true }
  end
end

class FailingPlainJob
  include Shoryuken::Worker
  shoryuken_options queue: 'failing_plain'

  def perform(*_args)
    raise 'Badaboom'
  end
end


spec/yabeda/shoryuken_spec.rb
# frozen_string_literal: true

RSpec.describe Yabeda::Shoryuken do
  let(:sample_body) { SecureRandom.uuid }

  it 'has a version number' do
    expect(Yabeda::Shoryuken::VERSION).not_to be nil
  end

  it 'configures middlewares' do
    expect(Shoryuken.client_middleware.entries).to include(have_attributes(klass: Yabeda::Shoryuken::ClientMiddleware))
  end

  describe 'plain Shoryuken jobs' do
    it 'counts enqueues' do
      Yabeda.shoryuken.messages_enqueued_total.values.clear # This is a hack

      Shoryuken::Client.queues('default').send_message(message_body: sample_body)

      expect(Yabeda.shoryuken.messages_enqueued_total.values).to include(
        { queue: 'default' } => 1,
      )
    end

    it 'measures runtime' do
      Yabeda.shoryuken.jobs_executed_total.values.clear   # This is a hack
      Yabeda.shoryuken.jobs_success_total.values.clear    # This is a hack
      Yabeda.shoryuken.jobs_failed_total.values.clear     # This is a hack
      Yabeda.shoryuken.job_runtime.values.clear           # This is a hack also

      SamplePlainJob.perform_async(sample_body)
      SamplePlainJob.perform_async(sample_body)
      begin
        FailingPlainJob.perform_async(sample_body)
      rescue StandardError
        nil
      end

      expect(Yabeda.shoryuken.jobs_executed_total.values).to eq(
        { queue: 'sample_plain', worker: 'SamplePlainJob' } => 2,
        { queue: 'failing_plain', worker: 'FailingPlainJob' } => 1,
      )
      expect(Yabeda.shoryuken.jobs_success_total.values).to eq(
        { queue: 'sample_plain', worker: 'SamplePlainJob' } => 2,
      )
      expect(Yabeda.shoryuken.jobs_failed_total.values).to eq(
        { queue: 'failing_plain', worker: 'FailingPlainJob' } => 1,
      )
      expect(Yabeda.shoryuken.job_runtime.values).to include(
        { queue: 'sample_plain', worker: 'SamplePlainJob' } => kind_of(Numeric),
        { queue: 'failing_plain', worker: 'FailingPlainJob' } => kind_of(Numeric),
      )
    end
  end

  describe '#yabeda_tags worker method' do
    it 'uses custom labels for both shoryuken and application metrics' do
      Yabeda.shoryuken.jobs_executed_total.values.clear   # This is a hack
      Yabeda.shoryuken.job_runtime.values.clear           # This is a hack also
      Yabeda.test.whatever.values.clear                 # And this

      begin
        SampleComplexJob.perform_async(sample_body)
      rescue StandardError
        nil
      end

      expect(Yabeda.shoryuken.jobs_executed_total.values).to eq(
        { queue: 'sample_complex', worker: 'SampleComplexJob', implicit: true } => 1,
      )
      expect(Yabeda.shoryuken.job_runtime.values).to include(
        { queue: 'sample_complex', worker: 'SampleComplexJob', implicit: true } => kind_of(Numeric),
      )
      expect(Yabeda.test.whatever.values).to include(
        { explicit: true, implicit: true } => 1,
      )
    end
  end

  describe 'collection of Shoryuken statistics' do
    it 'measures maximum runtime of currently running jobs' do
      Yabeda.shoryuken.running_job_runtime.values.clear # This is a hack
      described_class.jobs_started_at.clear

      workers = []
      workers.push(Thread.new { SampleLongRunningJob.perform_async(sample_body) })
      sleep 0.013 # Ruby can sleep less than requested
      workers.push(Thread.new { SampleLongRunningJob.perform_async(sample_body) })

      Yabeda.collectors.each(&:call)
      expect(Yabeda.shoryuken.running_job_runtime.values).to include(
        { queue: 'sample_long_running', worker: 'SampleLongRunningJob' } => (be >= 0.010),
      )

      sleep 0.013 # Ruby can sleep less than requested
      begin
        FailingPlainJob.perform_async(sample_body)
      rescue StandardError
        nil
      end
      Yabeda.collectors.each(&:call)

      expect(Yabeda.shoryuken.running_job_runtime.values).to include(
        { queue: 'sample_long_running', worker: 'SampleLongRunningJob' } => (be >= 0.020),
        { queue: 'failing_plain', worker: 'FailingPlainJob' } => 0,
      )

      # When all jobs are completed, metric should respond with zero
      workers.map(&:join)
      Yabeda.collectors.each(&:call)
      expect(Yabeda.shoryuken.running_job_runtime.values).to include(
        { queue: 'sample_long_running', worker: 'SampleLongRunningJob' } => 0,
        { queue: 'failing_plain', worker: 'FailingPlainJob' } => 0,
      )
    end
  end
end


lib/yabeda/shoryuken/config.rb
# frozen_string_literal: true

require 'anyway'

module Yabeda
  module Shoryuken
    # Yabeda Shoryuken config
    class Config < ::Anyway::Config
      config_name :yabeda_sidekiq

      # By default all sidekiq worker processes (servers) collects global metrics about whole Sidekiq installation.
      # Client processes (everything else that is not Sidekiq worker) by default doesn't.
      # With this config you can override this behavior:
      #  - force disable if you don't want multiple Sidekiq workers to report the same numbers (that causes excess load to both Redis and monitoring)
      #  - force enable if you want non-Sidekiq process to collect them (like dedicated metric exporter process)
      attr_config collect_cluster_metrics: ::Shoryuken.server?

      # Declare metrics that are only tracked inside worker process even outside them
      attr_config declare_process_metrics: ::Shoryuken.server?
    end
  end
end


lib/yabeda/shoryuken/server_middleware.rb
# frozen_string_literal: true

module Yabeda
  module Shoryuken
    # Shoryuken worker middleware
    class ServerMiddleware
      # rubocop: disable Metrics/AbcSize, Style/ExplicitBlockArgument
      def call(worker, queue, sqs_msg, body)
        jid = SecureRandom.uuid

        custom_tags = Yabeda::Shoryuken.custom_tags(worker, sqs_msg).to_h
        labels = Yabeda::Shoryuken.labelize(worker.class, sqs_msg, queue, body).merge(custom_tags)
        start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        begin
          Yabeda::Shoryuken.jobs_started_at[labels][jid] = start
          Yabeda.with_tags(**custom_tags) do
            yield
          end

          Yabeda.shoryuken_jobs_success_total.increment(labels)
        rescue Exception # rubocop: disable Lint/RescueException
          Yabeda.shoryuken_jobs_failed_total.increment(labels)
          raise
        ensure
          Yabeda.shoryuken_job_runtime.measure(labels, elapsed(start))
          Yabeda.shoryuken_jobs_executed_total.increment(labels)
          Yabeda::Shoryuken.jobs_started_at[labels].delete(jid)
        end
      end
      # rubocop: enable Metrics/AbcSize, Style/ExplicitBlockArgument

      private

      def elapsed(start)
        (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start).round(3)
      end
    end
  end
end


lib/yabeda/shoryuken.rb
# frozen_string_literal: true

require 'shoryuken'

require 'yabeda'
require 'yabeda/shoryuken/version'
require 'yabeda/shoryuken/client_middleware'
require 'yabeda/shoryuken/server_middleware'
require 'yabeda/shoryuken/config'

module Yabeda
  # Yabeda Shoryuken integration
  module Shoryuken
    LONG_RUNNING_JOB_RUNTIME_BUCKETS = [
      0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, # standard (from Prometheus)
      30, 60, 120, 300, 1800, 3600, 21_600, # Sidekiq tasks may be very long-running
    ].freeze

    def self.config
      @config ||= Config.new
    end

    Yabeda.configure do
      config = ::Yabeda::Shoryuken.config

      group :shoryuken

      counter :messages_enqueued_total, tags: %i[queue],
                                        comment: 'A counter of the total number of message shoryuken enqueued.'

      if config.declare_process_metrics # defaults to +::Shoryuken.server?+
        counter   :jobs_executed_total,  tags: %i[queue worker],
                                         comment: 'A counter of the total number of jobs shoryuken executed.'
        counter   :jobs_success_total,   tags: %i[queue worker],
                                         comment: 'A counter of the total number of jobs successfully processed by shoryuken.'
        counter   :jobs_failed_total,    tags: %i[queue worker],
                                         comment: 'A counter of the total number of jobs failed in shoryuken.'

        gauge     :running_job_runtime,  tags: %i[queue worker], aggregation: :max, unit: :seconds,
                                         comment: 'How long currently running jobs are running (useful for detection of hung jobs)'

        histogram :job_runtime, comment: 'A histogram of the job execution time.',
                                unit: :seconds, per: :job,
                                tags: %i[queue worker],
                                buckets: LONG_RUNNING_JOB_RUNTIME_BUCKETS
      end

      collect do
        Yabeda::Shoryuken.track_max_job_runtime if ::Shoryuken.server?
      end
    end

    ::Shoryuken.configure_server do |config|
      config.server_middleware do |chain|
        chain.add ServerMiddleware
      end
      config.client_middleware do |chain|
        chain.add ClientMiddleware
      end
    end

    ::Shoryuken.configure_client do |config|
      config.client_middleware do |chain|
        chain.add ClientMiddleware
      end
    end

    class << self
      def labelize(worker_class, sqs_msg, queue, body = nil)
        { queue: queue, worker: worker_name(worker_class, sqs_msg, body) }
      end

      def worker_name(worker_class, sqs_msg, body = nil)
        if ::Shoryuken.active_job? \
          && !sqs_msg.is_a?(Array) \
          && sqs_msg.message_attributes \
          && sqs_msg.message_attributes['shoryuken_class'] \
          && sqs_msg.message_attributes['shoryuken_class'][:string_value] \
          == ActiveJob::QueueAdapters::ShoryukenAdapter::JobWrapper.to_s \
          && body

          "ActiveJob/#{body['job_class']}"
        else
          worker_class.to_s
        end
      end

      def custom_tags(worker, sqs_msg)
        return {} unless worker.respond_to?(:yabeda_tags)

        worker.method(:yabeda_tags).arity.zero? ? worker.yabeda_tags : worker.yabeda_tags(*sqs_msg.attributes)
      end

      # Hash of hashes containing all currently running jobs' start timestamps
      # to calculate maximum durations of currently running not yet completed jobs
      # { { queue: "default", worker: "SomeJob" } => { "jid1" => 100500, "jid2" => 424242 } }
      attr_accessor :jobs_started_at

      def track_max_job_runtime
        now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        ::Yabeda::Shoryuken.jobs_started_at.each do |labels, jobs|
          oldest_job_started_at = jobs.values.min
          oldest_job_duration = oldest_job_started_at ? (now - oldest_job_started_at).round(3) : 0
          Yabeda.shoryuken.running_job_runtime.set(labels, oldest_job_duration)
        end
      end
    end

    self.jobs_started_at = Concurrent::Map.new { |hash, key| hash[key] = Concurrent::Map.new }
  end
end
